File: src/keepsake/store/pool.py
"""Connection pooling and tenant scoping. The GUC is set here and nowhere else."""

from collections.abc import Iterator
from contextlib import contextmanager
from uuid import UUID

import psycopg
from psycopg_pool import ConnectionPool

from keepsake.store import ADMIN_GUC, POOL_SIZE, SCHEMA, TENANT_GUC, validated_schema

# search_path is an ordinary GUC, so it travels as a parameter rather than as DDL.
_SET_SEARCH_PATH = "SELECT set_config('search_path', %s, true)"

# The two policies OR into one expression, and tenant_isolation casts TENANT_GUC to
# uuid whichever way that OR is planned: Postgres does not promise short-circuiting
# and says outright not to rely on it to avoid an error. An unset GUC reads NULL or
# '', and ''::uuid raises. This value casts cleanly, matches nothing, and so leaves
# admin_read to supply the true.
_NIL_TENANT = "00000000-0000-0000-0000-000000000000"


class Store:
    def __init__(self, dsn: str, schema: str = SCHEMA) -> None:
        self._search_path = f"{validated_schema(schema)}, pg_catalog"
        # max_size, not just the default min_size: psycopg's pool otherwise never grows
        # past four connections, which caps the whole pod at four concurrent writes.
        self._pool = ConnectionPool(
            dsn,
            open=True,
            min_size=1,
            max_size=POOL_SIZE,
            # Unchecked by default, and a pooled connection does not notice the server
            # going away: every restart or failover then hands each waiting agent one
            # dead connection and "terminating connection due to administrator command",
            # as a transport failure it cannot act on. The check is one round trip and
            # replaces the connection instead.
            check=ConnectionPool.check_connection,
            # The reconnect delay doubles — 1s, 2s, 4s — and is bounded only by this
            # window. At the 300s default a replica that failed for a minute sleeps the
            # next minute too, so it is still refusing writes long after the database
            # came back. Giving up sooner is what shortens that: the pool drops the
            # attempt and the next request starts a fresh one.
            reconnect_timeout=15.0,
            kwargs={"autocommit": False},
        )

    def close(self) -> None:
        self._pool.close()

    def healthy(self, timeout: float = 2.0) -> bool:
        """Whether the pool can hand out a working connection right now.

        Answers rather than raises: this drives a readiness probe, and the one thing
        it must never do is fail to produce a verdict. The timeout is short on purpose
        — the probe reports "not ready" far sooner than the pool's own 30s wait.
        """
        try:
            with self._pool.connection(timeout=timeout) as conn:
                conn.execute("SELECT 1")
        except Exception:  # noqa: BLE001 - a probe reports, it does not propagate
            return False
        return True

    @contextmanager
    def raw(self) -> Iterator[psycopg.Connection]:
        """A connection with no tenant scope. For startup checks only."""
        with self._pool.connection() as conn, conn.transaction():
            # Enforces the docstring rather than advertising it: the pool commits on
            # clean exit, so an unscoped connection is otherwise a usable write path.
            conn.execute("SET TRANSACTION READ ONLY")
            conn.execute(_SET_SEARCH_PATH, (self._search_path,))
            yield conn

    @contextmanager
    def scope(self, tenant_id: UUID) -> Iterator[psycopg.Connection]:
        """Yield a connection scoped to `tenant_id` for the life of one transaction."""
        with self._pool.connection() as conn, conn.transaction():
            # set_config's is_local argument is SET LOCAL and, unlike SET, takes a
            # parameter. A session-scoped value would outlive the transaction and be
            # inherited by whoever next takes this connection from the pool. One
            # statement rather than two: this runs on every read the server serves.
            conn.execute(
                "SELECT set_config('search_path', %s, true),"
                "       set_config(%s, %s, true)",
                (self._search_path, TENANT_GUC, str(tenant_id)),
            )
            yield conn

    @contextmanager
    def admin_scope(self) -> Iterator[psycopg.Connection]:
        """Yield a connection that reads every tenant, for the admin console only.

        Read-only, and the policy behind it is FOR SELECT: an admin has no write
        path into a tenant it did not name. The read is gated far more weakly —
        okf.admin is self-asserted, so anything holding the app DSN can set it and
        read every tenant, with or without this method. That follows from gating on
        a GUC with no TO clause, which is what keeps the policy working in
        postgres.mode: existing, where the role has a name we do not know.
        Authentication therefore happens above this method and never inside it.
        """
        with self._pool.connection() as conn, conn.transaction():
            conn.execute("SET TRANSACTION READ ONLY")
            conn.execute(
                "SELECT set_config('search_path', %s, true),"
                "       set_config(%s, 'on', true),"
                "       set_config(%s, %s, true)",
                (self._search_path, ADMIN_GUC, TENANT_GUC, _NIL_TENANT),
            )
            yield conn

File: internal/store/pool.go
package store

import (
	"context"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

// acquireTimeout bounds only the wait to acquire a pooled connection, as psycopg_pool's does.
var acquireTimeout = 30 * time.Second

// SetAcquireTimeout overrides acquireTimeout for a test and returns a restorer.
func SetAcquireTimeout(d time.Duration) func() {
	orig := acquireTimeout
	acquireTimeout = d
	return func() { acquireTimeout = orig }
}

// nilTenant gives admin_read's uuid cast of an unset TenantGUC a value that matches no tenant.
var nilTenant = uuid.Nil.String()

// Store owns the pool and sets the tenant GUC, ported from 8f2af2e:src/keepsake/store/pool.py.
type Store struct {
	pool       *pgxpool.Pool
	searchPath string
}

// Open creates the pool for dsn, scoped to schema.
func Open(ctx context.Context, dsn, schema string) (*Store, error) {
	schema, err := ValidatedSchema(schema)
	if err != nil {
		return nil, err
	}
	poolSize, err := PoolSize()
	if err != nil {
		return nil, err
	}

	cfg, err := pgxpool.ParseConfig(dsn)
	if err != nil {
		return nil, err
	}
	cfg.MaxConns = int32(poolSize)
	cfg.MinConns = 1
	// Every acquire, as pool.py's check_connection: pgxpool replaces a dead connection instead of failing the caller.
	cfg.ShouldPing = func(ctx context.Context, _ pgxpool.ShouldPingParams) bool { return true }

	pool, err := pgxpool.NewWithConfig(ctx, cfg)
	if err != nil {
		return nil, err
	}
	return &Store{pool: pool, searchPath: schema + ", pg_catalog"}, nil
}

// OpenVerified opens the store and refuses it unless Verify passes.
func OpenVerified(ctx context.Context, dsn, schema string) (*Store, error) {
	s, err := Open(ctx, dsn, schema)
	if err != nil {
		return nil, err
	}
	if err := Verify(ctx, s, schema); err != nil {
		s.Close()
		return nil, err
	}
	return s, nil
}

// Stat snapshots the pool's connection counts.
func (s *Store) Stat() *pgxpool.Stat { return s.pool.Stat() }

// Close releases the pool's connections.
func (s *Store) Close() { s.pool.Close() }

// Healthy reports whether the pool can hand out a working connection; it never raises.
func (s *Store) Healthy(ctx context.Context) bool {
	ctx, cancel := context.WithTimeout(ctx, 2*time.Second)
	defer cancel()
	// The acquire pings, and replaces a dead connection, so it is the whole check.
	conn, err := s.pool.Acquire(ctx)
	if err != nil {
		return false
	}
	conn.Release()
	return true
}

// Raw yields a read-only connection with no tenant scope, for startup checks only.
func (s *Store) Raw(ctx context.Context, fn func(pgx.Tx) error) error {
	return s.tx(ctx, fmt.Sprintf("BEGIN READ ONLY; SELECT set_config('search_path', '%s', true)", s.searchPath), fn)
}

// Scope yields a connection scoped to tenant for the life of one transaction.
func (s *Store) Scope(ctx context.Context, tenant uuid.UUID, fn func(pgx.Tx) error) error {
	// okf.admin off explicitly: a role or DSN default of 'on' would widen every read.
	return s.tx(ctx, fmt.Sprintf(
		"BEGIN; SELECT set_config('search_path', '%s', true), set_config('%s', 'off', true), set_config('%s', '%s', true)",
		s.searchPath, AdminGUC, TenantGUC, tenant), fn)
}

// AdminScope yields a read-only connection reading every tenant; authentication happens above it.
func (s *Store) AdminScope(ctx context.Context, fn func(pgx.Tx) error) error {
	return s.tx(ctx, fmt.Sprintf(
		"BEGIN READ ONLY; SELECT set_config('search_path', '%s', true), set_config('%s', 'on', true), set_config('%s', '%s', true)",
		s.searchPath, AdminGUC, TenantGUC, nilTenant), fn)
}

// tx runs fn in a transaction whose begin, built from validated values only, sets GUCs SET LOCAL in one round trip.
func (s *Store) tx(ctx context.Context, begin string, fn func(pgx.Tx) error) error {
	// The timeout covers only the acquire, as psycopg_pool's does, not the transaction.
	acquireCtx, cancel := context.WithTimeout(ctx, acquireTimeout)
	conn, err := s.pool.Acquire(acquireCtx)
	cancel()
	if err != nil {
		return &acquireError{err}
	}
	defer conn.Release()
	return pgx.BeginTxFunc(ctx, conn, pgx.TxOptions{BeginQuery: begin}, fn)
}

// acquireError marks an acquire failure, so IsUnavailable can tell it from fn's own deadline.
type acquireError struct{ err error }

func (e *acquireError) Error() string { return e.err.Error() }
func (e *acquireError) Unwrap() error { return e.err }
